swarm/network/pss.go

```diff
@@ -1,52 +1,384 @@
+		glog.V(logger.Detail).Infof("Found in block-cache: PSS-relay msg FROM %x TO %x", common.ByteLabel(self.Overlay.GetAddr().OverlayAddr()), common.ByteLabel(msg.To))
 		return nil
+		//return newPssError(pssError_BlockByCache)
 	}
 
-	ps.EachLivePeer(to, 255, func(p Peer, po int) bool {
-		err := p.Send(pssmsg)
+	self.addToFwdCache(digest)
+
+	// TODO:check integrity of message
+
+	sent := false
+
+	// send with kademlia
+	// find the closest peer to the recipient and attempt to send
+	self.Overlay.EachLivePeer(msg.To, 256, func(p Peer, po int) bool {
+		glog.V(logger.Debug).Infof("Attempting PSS-relay msg FROM %x TO %x THROUGH %x (%x)", common.ByteLabel(self.Overlay.GetAddr().OverlayAddr()), common.ByteLabel(msg.To), common.ByteLabel(p.OverlayAddr()), common.ByteLabel(p.UnderlayAddr()))
+		err := p.Send(msg)
 		if err != nil {
+			glog.V(logger.Warn).Infof("Attempting PSS-relay msg FROM %x TO %x THROUGH %x (%x) FAILED: %v", common.ByteLabel(self.Overlay.GetAddr().OverlayAddr()), common.ByteLabel(msg.To), common.ByteLabel(p.OverlayAddr()), common.ByteLabel(p.UnderlayAddr()), err)
 			return true
 		}
+		sent = true
 		return false
 	})
-	
+	if !sent {
+		return fmt.Errorf("PSS Was not able to send to any peers")
+	}
+
 	return nil
 }
+
+// we need the pss pointer here to access the final send
+type PssReadWriter struct {
+	*Pss
+	RecipientOAddr pot.Address
+	LastActive     time.Time
+	rw             chan p2p.Msg
+	ct             *protocols.CodeMap
+	topic          *PssTopic
+}
+
+// get incoming msg
+func (prw PssReadWriter) ReadMsg() (p2p.Msg, error) {
+	msg := <-prw.rw
+
+	glog.V(logger.Detail).Infof("pssrw readmsg %v", msg)
+
+	return msg, nil
+}
+
+func (prw PssReadWriter) WriteMsg(msg p2p.Msg) error {
+	glog.V(logger.Detail).Infof("pssrw writemsg %v", msg)
+	ifc, found := prw.ct.GetInterface(msg.Code)
+	if !found {
+		return fmt.Errorf("Writemsg couldn't find matching interface for code %d", msg.Code)
+	}
+	msg.Decode(ifc)
+
+	to := prw.RecipientOAddr.Bytes()
+
+	glog.V(logger.Detail).Infof("pssrw writemsg %v", msg)
+
+	pmsg, _ := MakeMsg(msg.Code, ifc)
+
+	return prw.Pss.Send(to, *prw.topic, pmsg)
+}
+
+// insert a incoming message on the p2p.VirtualPeer hack
+func (prw PssReadWriter) injectMsg(msg p2p.Msg) error {
+	glog.V(logger.Detail).Infof("pssrw injectmsg %v", msg)
+	prw.rw <- msg
+	return nil
+}
+
+// encapsulates all the complexity of the bridging of real-to-virtual protocol instantiations
+// TODO: needs review to see if it can be made leaner
+type PssProtocol struct {
+	*Pss
+	virtualProtocol *p2p.Protocol
+	topic           *PssTopic
+	ct              *protocols.CodeMap
+}
+
+func NewPssProtocol(pss *Pss, topic *PssTopic, ct *protocols.CodeMap, targetprotocol *p2p.Protocol) *PssProtocol {
+	pp := &PssProtocol{
+		Pss:             pss,
+		virtualProtocol: targetprotocol,
+		topic:           topic,
+		ct:              ct,
+	}
+	return pp
+}
+
+func (self *PssProtocol) GetHandler() func([]byte, *p2p.Peer, []byte) error {
```

Comment:
because it provides access to the convenience method that passes to p2p.Protocol handling, and thus needs to be available to the implementer.